File: p56_longestCommonSubsequence.cpp
// DP
#include <ioStream>

const int N_MAX = 1000, M_MAX = 1000;

int N, M;
char S[N_MAX], T[M_MAX];
int DP[N_MAX+1][M_MAX+1];

void solve(){

	for(int i=0; i<N; i++){
		for(int j=0; j<M; j++){
			if(S[i] == T[j]){
				DP[i+1][j+1] = DP[i][j]+1;
			} else{
				DP[i+1][j+1] = std::max(DP[i][j+1], DP[i+1][j]);
			}

		}
	}

	std::cout << DP[N][M] << '\n';
}
int main(){

	std::cin >> N >> M;
	std::cin >> S;
	std::cin >> T;

	solve();
}

File: p56_longestCommonSubsequence.py
n = int( input() )
m = int( input() )
s = list( input() )
t = list( input() )

dp = [[0 for i in range(m+1)] for j in range(n+1)]

for i in range(n):
    for j in range(m):
        if s[i]==t[j]:
            dp[i+1][j+1] = dp[i][j]+1
        else :
            dp[i+1][j+1] = max(dp[i+1][j], dp[i][j+1])

print(dp[n][m])